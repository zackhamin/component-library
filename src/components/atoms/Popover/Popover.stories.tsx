import { Button } from 'Atoms/Button';
import { Popover, PopoverProps } from './Popover';

export default {
	title: 'atoms/Popover',
	component: Popover,
};

export const PopoverStory = (args: PopoverProps) => (
	<>
		<Popover {...args} />
		<p>
			Lorem ipsum dolor sit amet consectetur, adipisicing elit. Laborum
			provident reprehenderit cupiditate assumenda quis recusandae tempore
			quisquam officiis, neque sunt delectus non ipsum pariatur quaerat at
			consequuntur perspiciatis alias est!
		</p>
	</>
);

PopoverStory.storyName = 'Popover';
PopoverStory.args = {
	title: 'My Popover',
	id: 'my-popover',
	children: (
		<p>
			Lorem ipsum dolor sit amet consectetur adipisicing elit. Exercitationem et
			doloremque neque illo itaque quisquam aperiam placeat tempore saepe
			tempora nesciunt minus mollitia recusandae magni, ipsam cum consequuntur
			ipsa atque.
		</p>
	),
	trigger: <Button>Trigger me!</Button>,
	width: '50rem',
	defaultOpen: true,
	gap: '0.8rem',
	align: 'left',
};
