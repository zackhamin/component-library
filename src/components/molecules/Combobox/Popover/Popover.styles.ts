import { colors } from 'Styles';
import styled from 'styled-components';

export const StyledPopover = styled.div`
	z-index: 1;
	position: absolute;
	width: 100%;
	top: 100%;
	border: 1px solid ${colors.lightGrey};
	border-radius: 4px;
	margin-top: 4px;
	box-shadow: rgba(0, 0, 0, 0.1) 0px 10px 15px -3px,
		rgba(0, 0, 0, 0.05) 0px 4px 6px -2px;
	background: ${colors.white};
`;

export const StyledPopoverHeading = styled.div`
	margin-bottom: 0.5rem;
`;
