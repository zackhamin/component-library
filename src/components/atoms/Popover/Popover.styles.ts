import { PopoverProps } from './Popover';
import styled from 'styled-components';
import { colors, visuallyHiddenStyles } from 'Styles';

const StyledWrapper = styled.div<Partial<PopoverProps>>`
	position: relative;
	display: inline-flex;
	flex-direction: column;
	align-items: ${({ align }) => {
		if (align === 'left') {
			return 'flex-start';
		}
		if (align === 'right') {
			return 'flex-end';
		}
		return 'center';
	}};
`;

const StyledPopover = styled.div<Partial<PopoverProps>>`
	position: absolute;
	top: 100%;
	background-color: ${colors.white};
	align-items: center;
	box-shadow: 0px 2px 6px rgba(51, 51, 51, 0.16);
	border-radius: 0.8rem;
	margin-top: ${({ gap }) => gap};
	padding: 1.6rem;
	min-width: 25rem;
	max-width: 90vw;
	width: ${({ width }) => width};
	z-index: 1;

	.visually-hidden {
		${visuallyHiddenStyles};
	}
`;

const StyledTrigger = styled.span`
	display: inline;
`;

export { StyledPopover, StyledTrigger, StyledWrapper };
