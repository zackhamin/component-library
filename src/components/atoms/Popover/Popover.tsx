import { Size } from '@Types/css';
import {
	cloneElement,
	FC,
	PropsWithChildren,
	ReactElement,
	useEffect,
	useRef,
	useState,
} from 'react';
import { StyledPopover, StyledTrigger, StyledWrapper } from './Popover.styles';

import classNames from 'classnames';
import { Heading } from 'Atoms/Heading';
import { useClickOutside } from 'Hooks/useClickOutside';
import { useOnKeyDown } from 'Hooks/useOnKeyDown';

export interface PopoverProps {
	id: string;
	/** The title of the popover (can be visually hidden with the `hideTitle` prop) */
	title: string;
	/** Visually hide the title of the popover and only keep it visible for screen readers */
	hideTitle?: boolean;
	/**
	 * A JSX element to be used as the trigger for the popover
	 * @default false
	 */
	trigger: ReactElement;
	/**
	 * The size of the popover
	 * @default 'auto'
	 */
	width?: Size | 'auto';
	/**
	 * The gap between the trigger and the popover
	 * @default '0.8rem'
	 */
	gap?: Size;
	/**
	 * Whether the popover should be open by default
	 * @default false
	 */
	defaultOpen?: boolean;
	/**
	 * Align the popover to the left, right or center of the trigger
	 * @default 'left'
	 */
	align?: 'left' | 'right' | 'center';
}

const Popover: FC<PropsWithChildren<PopoverProps>> = ({
	children,
	id,
	title,
	hideTitle = false,
	trigger,
	width = 'auto',
	gap = '0.8rem',
	defaultOpen = false,
	align = 'left',
}) => {
	const [isOpen, setIsOpen] = useState(defaultOpen);
	const [touched, setTouched] = useState(false);

	const outsideRef = useRef(null);
	const triggerRef = useRef<HTMLElement>(null);
	const popoverRef = useRef<HTMLDivElement>(null);

	const touch = () => {
		if (!touched) setTouched(true);
	};
	const close = () => {
		setIsOpen(false);
		touch();
	};
	const toggle = () => {
		setIsOpen((prev) => !prev);
		touch();
	};

	useOnKeyDown('Escape', close);
	useClickOutside(outsideRef, close);

	const triggerProps = {
		id: `${id}-trigger`,
		onClick: toggle,
		'aria-controls': id,
		'aria-expanded': isOpen,
		ref: triggerRef,
	};

	useEffect(() => {
		if (touched) {
			if (isOpen) {
				popoverRef.current?.focus();
			} else {
				triggerRef.current?.focus();
			}
		}
	}, [isOpen, touched]);

	return (
		<StyledWrapper ref={outsideRef} align={align}>
			<StyledTrigger>{cloneElement(trigger, triggerProps)}</StyledTrigger>
			{isOpen && (
				<StyledPopover
					ref={popoverRef}
					aria-labelledby="popover-title"
					width={width}
					tabIndex={-1}
					id={id}
					gap={gap}
				>
					<Heading
						id="popover-title"
						size="1.8rem"
						className={classNames({
							'visually-hidden': hideTitle,
						})}
					>
						{title}
					</Heading>
					{children}
				</StyledPopover>
			)}
		</StyledWrapper>
	);
};

export { Popover };
