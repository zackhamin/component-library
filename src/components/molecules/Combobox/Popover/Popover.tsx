import { StyledPopover } from './Popover.styles';
import React from 'react';
import { DismissButton, FocusScope, useOverlay } from 'react-aria';

interface PopoverProps {
	popoverRef?: React.RefObject<HTMLDivElement>;
	children: React.ReactNode;
	isOpen?: boolean;
	onClose: () => void;
}

export function Popover(props: PopoverProps) {
	const ref = React.useRef<HTMLDivElement>(null);
	const { popoverRef = ref, isOpen, onClose, children } = props;

	// Handle events that should cause the popup to close,
	// e.g. blur, clicking outside, or pressing the escape key.
	const { overlayProps } = useOverlay(
		{
			isOpen,
			onClose,
			shouldCloseOnBlur: true,
			isDismissable: true,
		},
		popoverRef
	);

	// Add a hidden <DismissButton> component at the end of the popover
	// to allow screen reader users to dismiss the popup easily.
	return (
		<FocusScope restoreFocus>
			<StyledPopover {...overlayProps} ref={popoverRef}>
				{children}
				<DismissButton onDismiss={onClose} />
			</StyledPopover>
		</FocusScope>
	);
}
