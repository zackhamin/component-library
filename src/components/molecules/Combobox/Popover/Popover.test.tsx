import { Popover } from './Popover';
import { fireEvent, render } from '@testing-library/react';

const defaultProps = {
	children: <p>Popover Content</p>,
	onClose: () => null,
};

describe('Popover', () => {
	it('opens popover on trigger click', () => {
		const { getAllByRole, getByText } = render(<Popover {...defaultProps} />);
		const trigger = getAllByRole('button')[0];
		fireEvent.click(trigger);

		const popover = getByText('Popover Content');
		expect(popover).toBeInTheDocument();
	});
	it('renders correctly', () => {
		const { container } = render(<Popover {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});
});
