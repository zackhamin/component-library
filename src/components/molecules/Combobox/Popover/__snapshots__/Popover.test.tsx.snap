// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders correctly 1`] = `
.c0 {
  z-index: 1;
  position: absolute;
  width: 100%;
  top: 100%;
  border: 1px solid #EEEEEF;
  border-radius: 4px;
  margin-top: 4px;
  box-shadow: rgba(0,0,0,0.1) 0px 10px 15px -3px, rgba(0,0,0,0.05) 0px 4px 6px -2px;
  background: #ffffff;
}

<div>
  <span
    data-focus-scope-start="true"
    hidden=""
  />
  <div
    class="c0"
  >
    <p>
      Popover Content
    </p>
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <button
        aria-label="Dismiss"
        id="react-aria-generated-id"
        tabindex="-1"
      />
    </div>
  </div>
  <span
    data-focus-scope-end="true"
    hidden=""
  />
</div>
`;
